workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt

```diff
@@ -0,0 +1,300 @@
+ * is required by Compose UI.
+ *
+ * This class is designed to support a navigation container view that owns a
+ * a set of navigation "frames", where a frame is something that can be navigated
+ * to/from. A frame loosely consists of a root [View] and its
+ * [ViewTreeSavedStateRegistryOwner]. For example:
+ *
+ *  - a back stack container view will own an instance of [StateRegistryAggregator],
+ *   and use it to assign a [ViewTreeSavedStateRegistryOwner] for its top view.
+ *
+ *  - a container view managing a set of windows will own an instance of
+ *   [StateRegistryAggregator], and use it to assign a [ViewTreeSavedStateRegistryOwner]
+ *   to each dialog's content view.
+ *
+ * Note that a [ViewTreeSavedStateRegistryOwner] works _in parallel_ to a
+ * [ViewTreeLifecycleOwner][androidx.lifecycle.ViewTreeLifecycleOwner]. This class
+ * presumes that each managed child [View] has a [WorkflowLifecycleOwner] in place
+ * serving that role.
+ *
+ * [attachToParentRegistry] must be called when the container view is attached to a window,
+ * and passed the parent registry. [detachFromParentRegistry] must be called when the
+ * container view is detached.
+ *
+ * Call [installChildSavedStateRegistryControllerOn] to put a [ViewTreeSavedStateRegistryOwner]
+ * in place on each managed child view, _before it is attached to a window_. After that:
+ *
+ *  - call [saveChildSavedStateRegistryOwner] if the child is removed from service but may
+ *   be restored before [detachFromParentRegistry] is called (as when a back stack
+ *   pushes and pops)
+ *
+ *  - call [prune] when views are permanently removed from service, taking care
+ *   to identify the set that remain active
+ */
+@WorkflowUiExperimentalApi
+public class StateRegistryAggregator {
+  /**
+   * Holds any states restored from the parent registry, as well as any states saved
+   * after that time via [saveChildSavedStateRegistryOwner].
+   *
+   * Will be null until we are restored from the parent registry. After being restored,
+   * it will never be re-assigned again.
+   */
+  private var states: MutableMap<String, Bundle>? = null
+
+  private val isRestored get() = states != null
+
+  /** Memoize the registry owner passed to [attachToParentRegistry] so it can be detached later. */
+  private var parentRegistryOwner: SavedStateRegistryOwner? = null
+  private var parentKey: String? = null
+
+  /**
+   * The set of [KeyedStateRegistryOwner] instances created by
+   * [installChildSavedStateRegistryControllerOn], which have not
+   * yet been retired via [saveChildSavedStateRegistryOwner] or [prune].
+   */
+  private val children = mutableMapOf<String, KeyedStateRegistryOwner>()
+
+  /**
+   * Used to observe the parent registry's lifecycle to know when it becomes `CREATED`
+   * and is ready for us to restore ourselves. This observer is only registered between
+   * calls to [attachToParentRegistry] and [detachFromParentRegistry], and will only
```

Comment:
Is this fragile in the sense that if we called `setContentView` from _within_ `onCreate()` which would attach the view to the window, which would `attachToParentRegistry` we could miss the `ON_CREATE`?

I'm trying to think through how this works on config change and process restoration ... why is this working? Are we not calling `setCOntentView` via `onCreate`? Is this installed in `onCreate` but the 'ON_CREATE` lifecycle event happens after `onCreate` is complete?